applications/CompressiblePotentialFlowApplication/custom_strategies/transonic_residualbased_incrementalupdate_static_scheme.h

```diff
@@ -0,0 +1,343 @@
+     */
+    explicit TransonicResidualBasedIncrementalUpdateStaticScheme(Parameters ThisParameters) : BaseType()
+    {
+        // Validate and assign defaults
+        ThisParameters = this->ValidateAndAssignParameters(ThisParameters, this->GetDefaultParameters());
+        this->AssignSettings(ThisParameters);
+    }
+
+
+    /** Destructor.
+    */
+    ~TransonicResidualBasedIncrementalUpdateStaticScheme() override {}
+
+    ///@}
+    ///@name Operators
+    ///@{
+
+    ///@}
+    ///@name Operations
+    ///@{
+
+    /**
+     * @brief Create method
+     * @param ThisParameters The configuration parameters
+     */
+    typename BaseSchemeType::Pointer Create(Parameters ThisParameters) const override
+    {
+        return Kratos::make_shared<ClassType>(ThisParameters);
+    }
+
+    /**
+     * @brief This is the place to initialize the Scheme.
+     * @details This is intended to be called just once when the strategy is initialized
+     * @param rModelPart The model part of the problem to solve
+     */
+    void Initialize(ModelPart& rModelPart) override
+    {
+        KRATOS_TRY
+
+        rModelPart.GetProcessInfo()[CRITICAL_MACH]          = mInitialCriticalMach;
+        rModelPart.GetProcessInfo()[UPWIND_FACTOR_CONSTANT] = mInitialUpwindFactorConstant;
+        rModelPart.GetProcessInfo()[MACH_LIMIT]             = std::sqrt(mMachNumberSquaredLimit);
+
+        BaseType::SetSchemeIsInitialized(true);
+        KRATOS_CATCH("")
+    }
+
+    /**
+     * @brief Function to be called when it is needed to initialize an iteration. It is designed to be called at the beginning of each non linear iteration
+     * @note Take care: the elemental function with the same name is NOT called here.
+     * @warning Must be defined in derived classes
+     * @details The function is called in the builder for memory efficiency
+     * @param rModelPart The model part of the problem to solve
+     * @param A LHS matrix
+     * @param Dx Incremental update of primary variables
+     * @param b RHS Vector
+     */
+    void InitializeNonLinIteration(
+        ModelPart& rModelPart,
+        TSystemMatrixType& A,
+        TSystemVectorType& Dx,
+        TSystemVectorType& b
+        ) override
+    {
+        KRATOS_TRY
+
+        // Update the upwind factor constant and critical mach with the user-defined values
+        if (rModelPart.GetProcessInfo()[CONVERGENCE_RATIO] <= mUpdateRelativeResidualNorm &&
+            rModelPart.GetProcessInfo()[NL_ITERATION_NUMBER] > 1                          &&
+            mUpdatedValues == false){
+
+            rModelPart.GetProcessInfo()[CRITICAL_MACH]          = mTargetCriticalMach;
+            rModelPart.GetProcessInfo()[UPWIND_FACTOR_CONSTANT] = mTargetUpwindFactorConstant;
+
+            mUpdatedValues = true;
+        }
+
+        // Initialize non-linear iteration for all of the elements, conditions and constraints
+        EntitiesUtilities::InitializeNonLinearIterationAllEntities(rModelPart);
```

Comment:
Same comment, shouldn't we call the base class `InitializeNonLinIteration` instead?

Reply:
yes, the same as before